sotodlib/site_pipeline/update_smurf_caldbs.py

```diff
@@ -109,15 +124,289 @@ def smurf_detset_info(config:str | dict,
             })
         write_dataset(
             det_rs, 
-            config['archive']['h5file'], 
+            config['archive']['detset']['h5file'], 
             ts.name, 
             overwrite,
         )
         # add new entries to database
         if ts.name not in added_detsets:
             db_data = {'dets:detset': ts.name,
                     'dataset': ts.name}
-            db.add_entry(db_data, config['archive']['h5file'])
+            db.add_entry(db_data, config['archive']['detset']['h5file'])
+
+
+def get_cal_obsids(ctx, obs_id, cal_type):
+    """
+    Returns set of obs-ids corresponding to the most recent calibration
+    operations for a given obsid.
+
+    Args
+    ------
+    ctx: core.Context
+        Context object
+    obs_id: str
+        obs_id for which you want to get relevant calibration info
+    cal_type: str
+        Calibration subtype to use in the obsdb query. For example: 'iv' or
+        'bias_steps'.
+
+    Returns
+    ----------
+        obs_ids: dict
+            Dict of obs_ids for each detset in specified operation
+    """
+    obs = ctx.obsdb.query(f"obs_id == '{obs_id}'")[0]
+    detsets = ctx.obsfiledb.get_detsets(obs_id)
+    min_ct = obs['start_time'] - 3600*24*7
+    cal_all = ctx.obsdb.query(
+        f"""
+        start_time <= {obs['start_time']} and subtype=='{cal_type}'
+        and start_time > {min_ct}
+        """, sort=['start_time']
+    )[::-1]
+
+    obs_ids = {
+        ds: None for ds in detsets
+    }
+
+    for o in cal_all:
+        dsets = ctx.obsfiledb.get_files(o['obs_id']).keys()
+        for ds in dsets:
+            if ds in obs_ids:
+                if obs_ids[ds] is None:
+                    obs_ids[ds] = o['obs_id']
+
+    return obs_ids
+
+# Dtype for calibration set
+cal_dtype = [
+    ('dets:readout_id', '<U40'),
+    ('r_tes', float),
+    ('r_frac', float),
+    ('p_bias', float),
+    ('s_i', float),
+    ('phase_to_pW', float),
+    ('bg', int),
+    ('polarity', int),
+    ('r_n', float),
+    ('p_sat', float),
+]
+
+@dataclass
+class CalInfo:
+    # Fields must be ordered like cal_dtype!
+    readout_id: str = ''
+
+    # From bias steps
+    r_tes: float = np.nan   # Ohm
+    r_frac: float = np.nan
+    p_bias: float = np.nan  # J
+    s_i: float = np.nan     # 1/V
+    phase_to_pW: float = np.nan  # pW/rad
+
+    # From IV
+    bg: int = -1
+    polarity: int = 1
+    r_n: float = np.nan  
+    p_sat: float = np.nan   # J
+
+
+def _load_smurf_npy(ctx, obs_id, substr):
```

Comment:
Could we move this function into `sotodlib.io.load_book`? This also seems like a function that other folks might want to use / something that's more general than site_pipeline